---
title: "Coral_FW_Gene_Expression"
author: "Elsa Brenner, Erica Sun, Nick Peoples"
date: "3/24/2021"
output: html_document
---
set the working directory
```{r}
setwd("YOUR FILE PATH")
```
## Install required packages
```{r, eval=FALSE}
BiocManager::install("DESeq2")
install.packages("backports")
install.packages("caTools")
BiocManager::install("affycoretools")
BiocManager::install("arrayQualityMetrics")
install.packages("pheatmap")
install.packages("tidyverse")
```

## Conduct array quality metrics to detect and remove outliers
Load the required packages.
```{r}
library(DESeq2)
library(affycoretools)
library(arrayQualityMetrics)
library(genefilter)
library(Biobase)
```

### Read in counts. 
allcounts_Harvey_fav.txt file is from BI586 GitHub. To make a count data .txt, follow the taqseq processing protocol.
```{r}
countData <- read.table("allcounts_Harvey_fav.txt")
head(countData)
length(countData[,1])
# set names for countData
names(countData)=c("fav_recoveryA",	"fav_recoveryB", "fav_recoveryC","fav_stressA", "fav_stressB", "fav_stressC")

#row.names(countData)=sub("", "isogroup", rownames(countData)) We don't need to modify the row names

head(countData)
```
### Look for outliers
Set the directory.
```{r}
setwd("~/Documents/GitHub/Assign2_Coral_FW/Outlier/")
v=setwd("~/Documents/GitHub/Assign2_Coral_FW/Outlier/")
```
Create a model using DESeq and normalizing the data.
```{r}
treat=c("fav_recovery",	"fav_recovery", "fav_recovery","fav_stress", "fav_stress", "fav_stress")
replicate = c("A", "B", "C", "A", "B", "C")
g = data.frame(cbind(treat, replicate))
g
colData= g

# create a model using DESeq too see how our design is varied by treatment 
dds=DESeqDataSetFromMatrix(countData=countData,
                           colData = g,
                           design = ~treat)

# normalizing the data
vsd.ge=assay(vst(dds)) 
rl=vst(dds)
e=ExpressionSet(assay(rl), AnnotatedDataFrame(as.data.frame(colData(rl))))
arrayQualityMetrics(e,outdir=v,intgroup=c("treat"),force=T)
```
Now we can go over to our Outliers directory and take a look at the index.html file. We didn't detect any outliers, but if there was an outlier you would have to remove them.

## DESeq Analysis
Since we don't have to run through the outlier chunk of code every time, let's read in counts data again. 
```{r}
countData <- read.table("allcounts_Harvey_fav.txt")
head(countData)
length(countData[,1])
names(countData)=c("fav_recoveryA",	"fav_recoveryB", "fav_recoveryC","fav_stressA", "fav_stressB", "fav_stressC")
head(countData)
```
Let's generate a table and visualize the total counts.
```{r}
totalCounts=colSums(countData)
totalCounts
barplot(totalCounts, col=c("coral", "coral", "coral", "red", "red", "red"), ylab="raw counts", main = "Total Counts of Recovery vs. Stress")
```
Looks like the total counts vary a little bit. Let's take a look at the counts in numbers.
```{r}
min(totalCounts) #458772
max(totalCounts)  #6596244
```
Create a model using DESeq and normalizing the data.
```{r}
treat=c("fav_recovery",	"fav_recovery", "fav_recovery","fav_stress", "fav_stress", "fav_stress")
replicate = c("A", "B", "C", "A", "B", "C")
g = data.frame(cbind(treat, replicate))
g
colData<- g

dds<-DESeqDataSetFromMatrix(countData=countData, colData=colData, design=~treat+replicate) #can only test for the main effects of site, pco2, temp

#one step DESeq
dds<-DESeq(dds)
```
Let's take a look at our DESeq result. 
```{r}
head(dds)
res<- results(dds)
```
Now let's plot the dispersions, which is looking at what DESeq is doing to our data. They do so by fitting our data into this curve.
```{r}
plotDispEsts(dds, main="Dispersion plot Corals")
```
Dispersion look like a hokey stick, which is good!
## Pairwise comparasion of stress vs. recovery
The order matters. Make sure that you put treatment first and control second.
```{r}
colData$treat<-factor(colData$treat, levels=c("fav_stress", "fav_recovery"))
resstress <- results(dds, contrast=c("treat","fav_stress", "fav_recovery"))
```
Let's take a look at the number of genes that meet the different thresholds. padj (p-adjusted) takes in account of the comparasions, rather than just looking at one object like p-value. If we're looking at a large amount of gene, it's easy to get a false positive just by chance. We have to do multiple test corrections to get the padj value.
```{r}
# number of differentially genes
#how many FDR < 10%?
table(resstress$padj<0.1)
# 0.1=46
# 0.05=28
# 0.01=19
summary(resstress)
```
There are 12392 samples removed because of low counts. Only 15 genes are down regulated under stress relative to recovery, and 31 genes are up regulated with the padj < 0.2.

Let's look at numbers of differential express genes that pass a threshold of 0.05. We find 28
```{r}
nrow(resstress[resstress$padj<0.05 & !is.na(resstress$padj),])
```
Now let's make a MA plot and save the result as a dataframe. 
```{r}
plotMA(resstress, main="Stress vs Recovery")
plotMA(resstress, main="Stress vs Recovery", ylim=c(-6,6)) # look at the same plot with differet ylim

results <- as.data.frame(resstress)
head(results) # somehow isogoup is doubled. Stil not sure how to fix this.
```
Get the number of genes that are up/down regulated. We find 31 up and 15 down
```{r}
nrow(resstress[resstress$padj<0.1 & resstress$log2FoldChange > 0 & !is.na(resstress$padj),])
nrow(resstress[resstress$padj<0.1 & resstress$log2FoldChange < 0 & !is.na(resstress$padj),])
```
Write a table of the results, and save to your directory as a .csv file
```{r}
write.table(resstress, file="stress_Harvey.txt", quote=F, sep="\t")

cd <- read.table("stress_Harvey.txt")
head(cd)
```

The following chunk creates a .csv file of the GO table for MWU (later)
```{r}
library(dplyr)
cd
go_input_stress = cd %>%
  tibble::rownames_to_column(var = "iso") %>%
  mutate(mutated_p = -log(pvalue)) %>%
  mutate(mutated_p_updown = ifelse(log2FoldChange < 0, mutated_p*-1, mutated_p*1)) %>%
  na.omit() %>%
  select(iso, mutated_p_updown)
head(go_input_stress)
colnames(go_input_stress) <- c("gene", "pval")
head(go_input_stress)
write.csv(go_input_stress, file="stress_GO.csv", quote=F, row.names=FALSE)
```

This next chunk gets the p values and creates a table of the rlogdata and p values
```{r}
valstress=cbind(resstress$pvalue, resstress$padj)
colnames(valstress)=c("pval.stress", "padj.stress")
length(valstress[,1])
table(complete.cases(valstress))

rlog=rlogTransformation(dds, blind=TRUE) 
rld=assay(rlog)
colnames(rld)=paste(colData$treat)
length(rld[,1])

rldpvals=cbind(rld,valstress)
dim(rldpvals)
# [1] 19937    8
table(complete.cases(rldpvals))
#FALSE  TRUE 
#15353  4584 

write.csv(rldpvals, "fav_Harvey_RLDandPVALS.csv", quote=F)

colnames(rld)=paste(colData$treat)
```

This chunk creates an improved sample distance heatmap of overall gene expression differences between the stress and recovery groups
```{r}
rldpvals <- read.csv(file="fav_Harvey_RLDandPVALS.csv", row.names=1)
head(rldpvals)

rld=rldpvals[,1:6]
head(rld)

sampleDists <- dist(t(rld))
sampleDistMatrix <- as.matrix( sampleDists )
treat=c( "fav_recoveryA", "fav_recoveryB", "fav_recoveryC", "fav_stressA", "fav_stressB", "fav_stressC")
colnames(sampleDistMatrix)=paste(treat)
rownames(sampleDistMatrix)=paste(treat)

library("pheatmap")
heat.colors = colorRampPalette(rev(c("blue","yellow","red")),bias=0.3)(100)
pheatmap(sampleDistMatrix,color = heat.colors,cex=0.9,border_color=NA,cluster_rows=T,cluster_cols=T)
```

Conduct principal component analysis and create PCA biplot to show variation in gene expression
```{r}
library(vegan)
library(ggplot2)
library(ggrepel)
library(tidyverse)

rld_t=t(rld)
pca <- prcomp(rld_t,center = TRUE, scale = FALSE)
li <- pca$sdev^2 / sum(pca$sdev^2)
pc1v <- round(li[1] * 100, 1)
pc2v <- round(li[2] * 100, 1)
pca_s <- as.data.frame(pca$x)
pca_s <- pca_s[,c(1,2)]
pca_s$Samples = row.names(pca_s)
pca_s$treat=colData$treat

cbPalette <- c("darkgoldenrod2",  "darkolivegreen3", "dodgerblue3")
ggplot(pca_s, aes(PC1, PC2, color = treat, pch = treat)) +
  geom_point(size=3) +
  #  geom_text_repel(aes(label=Samples)) +
  scale_colour_manual(values=cbPalette)+
  theme_bw() +
  # geom_density2d(alpha=.5)+
  geom_polygon(alpha=.2)+
  xlab(paste0("PC1: ",pc1v,"% variance")) +
  ylab(paste0("PC2: ",pc2v,"% variance")) 
```

Test if the groupings differ significantly by their PC scores on the PC1 and PC2 axis. Adonis is better in this case than ANOVA/MANOVA because it uses both the centroids and squared deviation of each sample to that centroid, so that it can detect significant differences when the groups center around each other, but the variance of each group from the centroid is different.
```{r}
adonis(pca$x ~ treat, data = pca_s, method='eu', na.rm = TRUE) #No significance detected, p=0.3
```



Create a heatmap of genes differentially expressed between the recovery and treatment groups
```{r}
rldpvals <- read.csv(file="fav_Harvey_RLDandPVALS.csv", row.names=1)
head(rldpvals)
rld_site= rldpvals[,1:6]
head(rld_site)
gg=read.table("orb_fav_iso2gene.tab.txt",sep="\t", row.names=1)
head(gg)

nrow(rldpvals[rldpvals$padj.stress<0.01& !is.na(rldpvals$padj.stress),])

topnum= 100 # number of DEGS
head(rldpvals)
top100=head(rldpvals[order(rldpvals$padj.stress), ],topnum)
head(top100)
length(top100[,1])
summary(top100)

library(pheatmap)
head(top100)
p.val=0.1 # FDR cutoff
conds=top100[top100$padj.stress<=p.val & !is.na(top100$padj.stress),]
length(conds[,1])

exp=conds[,1:6] # change numbers to be your vsd data columns
means=apply(exp,1,mean) # means of rows
explc=exp-means # subtracting them
head(explc)

ccol=colorRampPalette(rev(c("red","chocolate1","#FEE090","grey10", "cyan3","cyan")))(100)
col0=colorRampPalette(rev(c("chocolate1","#FEE090","grey10", "cyan3","cyan")))(100)

pheatmap(explc,cluster_cols=T,scale="row",color=col0, show_rownames = F)
```


